api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java

```diff
@@ -87,7 +87,148 @@ public void onShutdown() {
 	 */
 	public Obs saveObs(Obs obs, String changeMessage) throws APIException {
-		if (null != obs && null != obs.getConcept() && obs.getConcept().isComplex()
+		if(obs == null){
+			throw new APIException("obs cannot be null!!", (Object[]) null); //TODO: change this to a message string!!
+		}
+
+		if(obs.getId() != null && changeMessage == null){
+			throw new APIException("Obs.error.ChangeMessage.required", (Object[]) null);
+		}
+
+		handleExistingObsWithComplexConcept(obs);
+
+		ensureRequirePrivilege(obs);
+
+		//Should allow updating a voided Obs, it seems to be pointless to restrict it,
+		//otherwise operations like merge patients won't be possible when to moving voided obs
+		if (obs.getObsId() == null || obs.getVoided()) {
+			return saveNewAndVoidedObs(obs,changeMessage);
+		} else if(!obs.isDirty()){
+			setPersonFromEncounter(obs);
+			return saveObsNotDirty(obs, changeMessage);
+		} else {
+			setPersonFromEncounter(obs);
+			return saveExistingObs(obs,changeMessage);
+		}
+	}
+
+	private void setPersonFromEncounter(Obs obs) {
+		Encounter encounter = obs.getEncounter();
+		if (encounter != null) {
+			obs.setPerson(encounter.getPatient());
+		}
+	}
+
+	private void voidExistingObs(Obs obs, String changeMessage, Obs newObs) {
+		// void out the original observation to keep it around for
+		// historical purposes
+		try {
+			Context.addProxyPrivilege(PrivilegeConstants.DELETE_OBS);
+
+			// fetch a clean copy of this obs from the database so that
+			// we don't write the changes to the database when we save
+			// the fact that the obs is now voided
+			Context.evictFromSession(obs);
+			obs = Context.getObsService().getObs(obs.getObsId());
+			//delete the previous file from the appdata/complex_obs folder
+			if (newObs.hasPreviousVersion() && newObs.getPreviousVersion().isComplex()) {
+				File previousFile = new AbstractHandler().getComplexDataFile(obs);
+				previousFile.delete();
+			}
+			// calling this via the service so that AOP hooks are called
+			Context.getObsService().voidObs(obs, changeMessage);
+
+		}
+		finally {
+			Context.removeProxyPrivilege(PrivilegeConstants.DELETE_OBS);
+		}
+	}
+
+	private Obs saveExistingObs(Obs obs, String changeMessage) {
+		// get a copy of the passed in obs and save it to the
+		// database. This allows us to create a new row and new obs_id
+		// this method doesn't copy the obs_id
+		Obs newObs = Obs.newInstance(obs);
+
+		// unset any voided properties on the new obs
+		newObs.setVoided(false);
+		newObs.setVoidReason(null);
+		newObs.setDateVoided(null);
+		newObs.setVoidedBy(null);
+		// unset the creation stats
+		newObs.setCreator(null);
+		newObs.setDateCreated(null);
+		newObs.setPreviousVersion(obs);
+
+		RequiredDataAdvice.recursivelyHandle(SaveHandler.class, newObs, changeMessage);
+
+		// save the new row to the database with the changes that
+		// have been made to it
+		dao.saveObs(newObs);
+
+		saveObsGroup(newObs,null);
+
+		voidExistingObs(obs, changeMessage, newObs);
```

Comment:
We are actually voiding the obs using obsService.voidObs() method [here](https://github.com/bharatak/openmrs-core/blob/854f1648099266f2ef3e60ca43d19343b49865fd/api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java#L139)